### PR TITLE
WIP: Private Messages search &API

### LINKED ISSUE #11140: Inbox / private messages / PMs should be searchable by Username
As reported by cTheDragons and confirmed by me: 

"Unable to search by @username in the Messages (Inbox) section. Only via the display name."

In my opinion, the ideal situation is that Inbox conversations should be searchable by both Usernames and Display Names.

I'll leave this as suggestion-discussion for a few days in case the staff want to comment and then mark it as help wanted.

### PR DESCRIPTION
**Pull request by negue**
This is a WIP to add the ability:
- [x] to search for a specific message / username / display (#11140) in all conversations
  - [x] reloads the conversations sidebar
  - [x] on the conversation selection it only lists the messages that contain the search string
- [x] ability to jump to the context
  - [x] resets the search mode, reloads the normal conversations
  - [x] reload the messages-lists to 10 messages before (including the selected message)
- [x] load newer messages 
- [x] jump to recent messages (reset to current behavior)
- [x] send a new message in search mode - also resets to the recent messages
- [x] conversations are now loaded in pages (10 per load) to reduce more server load

- fixes #11140

### CODE REVIEW

File: website/server/libs/inbox/conversation.methods.js [hunk 1]
```diff
@@ -41,14 +41,23 @@ async function usersMapByConversations (users) {
...
-export async function listConversations (owner) {
+export async function listConversations (owner, searchMessage = null) {
+  const matchQuery = {
+    ownerId: owner._id,
+  };
+
+  if (searchMessage) {
+    const searchRegex = { $regex: new RegExp(`${searchMessage}`, 'i') };
+
+    matchQuery.text = searchRegex;
```
Comment (paglias) [inline website/server/libs/inbox/conversation.methods.js]: will this require a new index?

File: website/server/libs/inbox/index.js [hunk 2]
```diff
@@ -80,6 +77,59 @@ export async function getUserInbox (user, options = {
...
+  if (options.conversation) {
+    findObj.uuid = options.conversation;
+  }
+
+  if (options.beforeTimestamp) {
+    findObj.timestamp = { $lte: options.beforeTimestamp };
+  }
+
+  if (options.afterTimestamp) {
+    findObj.timestamp = Object.assign(findObj.timestamp || {},
```
Comment (paglias) [inline website/server/libs/inbox/index.js:106]: index?

**Review by paglias — CHANGES_REQUESTED** [review]
couple of comments about indexes so we won't forget

File: website/client/src/components/messages/messageList.vue [hunk 3]
```diff
@@ -74,6 +64,24 @@
         />
       </div>
     </div>
+    <div class="row loadmore">
```
Comment (negue) [inline website/client/src/components/messages/messageList.vue:69]: extract as a component? (same as load earlier messages)

File: website/client/src/pages/private-messages.vue [hunk 4]
```diff
@@ -555,9 +558,10 @@
 <script>
 import Vue from 'vue';
 import moment from 'moment';
-import filter from 'lodash/filter';
+// import filter from 'lodash/filter';
```
Comment (negue) [inline website/client/src/pages/private-messages.vue:615]: clean up

File: website/client/src/pages/private-messages.vue [hunk 5]
```diff
@@ -722,15 +748,9 @@ export default {
...
-
-      const filtered = subscribeToUpdate && !this.search
-        ? this.conversations
 
-        /* eslint-disable max-len */
-        : filter(this.conversations, conversation => conversation.name.toLowerCase().indexOf(this.search.toLowerCase()) !== -1);
-
-      const ordered = orderBy(filtered, [o => moment(o.date).toDate()], ['desc']);
+      const ordered = orderBy(this.conversations,
+        [o => moment(o.date).toDate()], ['desc']);
```
Comment (negue) [inline website/client/src/pages/private-messages.vue:809]: parsing string to moment to date really needed? using the normal string to order would probably sort the same?

File: website/server/libs/inbox/shared.methods.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+export function createSearchParams (searchText) {
+  const searchRegex = { $regex: new RegExp(`${searchText}`, 'i') };
+
+  return {
+    $or: [
```
Comment (negue) [inline website/server/libs/inbox/shared.methods.js:5]: added user/username , I don't know if these have an index already?

**Pull request closed without merging**